File: src/parser/lettersParser.cpp
#include "parser/lettersParser.hpp"

using namespace Parser;

State* LettersParser::Run(State* state) {
    if (state->IsError) { 
        return state; 
    }

    if (state->Input.empty() || (unsigned int)state->Input.size() < (unsigned int)state->Index + 1) {
        return setErrorState(state, "LettersParser: got unexpected end of input");
    }

    auto toCheck = state->Input.substr(state->Index);
    auto output = std::regex_match(state->Input.c_str(), std::regex(regex));

    if (!output) {
        return setErrorState(state, "LettersParser: couldn't match letters at index '" + std::to_string(state->Index) + "'");
    }

    state->Result = toCheck;
    state->Index++;
    state->IsError = false;
    state->Error = "";
    return state;
}

File: cmd/sealasm/parser/letters_parser.py
import re

from parser.base_parser import BaseParser
from parser.state import State


class LettersParser(BaseParser):
    "used to check a given string only contains letters"

    def __init__(self):
        self._regex: Pattern[AnyStr] = re.compile("^[A-Za-z]+")
        super().__init__()

    def run(self, state: State) -> State:
        if state.is_error:
            return state

        if not len(state.source) or len(state.source) < state.index + 1:
            return self._set_error_state(state, "LettersParser: got unexpected end of input")

        source = state.source[state.index:]
        output = self._regex.match(source)

        if output:
            state.result = output.string[output.start():output.end()]
            state.index = output.end()
            state.is_error = False
            state.error = ""
        else:
            state.error = f"LettersParser: couldn't match digits at index '{state.index}'"
            state.is_error = True

        return State(state=state)
